pkg/credentialprovider/apis/registrycredentials/types.go

```diff
@@ -0,0 +1,115 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registrycredentials
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type RegistryCredentialConfig struct {
+	metav1.TypeMeta `json:",inline"`
+
+	Providers []RegistryCredentialProvider `json:"providers"`
+}
+
+// RegistryCredentialProvider is used by the kubelet container runtime to match the
+// image property string (from the container spec) with exec-based credential provider
+// plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+
+	// ImageMatchers is a list of strings used to match against the image property
+	// (sometimes called "registry path") to determine which images to provide
+	// credentials for.  If one of the strings matches the image property, then the
+	// RegistryCredentialProvider will be used by kubelet to provide credentials
+	// for the image pull.
+
+	// The image property of a container supports the same syntax as the docker
+	// command does, including private registries and tags. A registry path is
+	// similar to a URL, but does not contain a protocol specifier (https://).
+	//
+	// Each ImageMatcher string is a pattern which can optionally contain
+	// a port and a path, similar to the image spec.  Globs can be used in the
+	// hostname (but not the port or the path).
+	//
+	// Globs are supported as subdomains (*.k8s.io) or (k8s.*.io), and
+	// top-level-domains (k8s.*).  Matching partial subdomains is also supported
+	// (app*.k8s.io).  Each glob can only match a single subdomain segment, so
+	// *.io does not match *.k8s.io.
+	//
+	// The image property matches when it has the same number of parts as the
+	// ImageMatcher string, and each part matches.  Additionally the path of
+	// ImageMatcher must be a prefix of the target URL. If the ImageMatcher
+	// contains a port, then the port must match as well.
+	ImageMatchers []string `json:"imageMatchers"`
+
+	// Exec specifies a custom exec-based plugin.  This type is defined in
+	Exec ExecConfig `json:"exec"`
+}
+
+// ExecConfig specifies a command to provide client credentials. The command is exec'd
+// and outputs structured stdout holding credentials.
+//
+// See the client.authentiction.k8s.io API group for specifications of the exact input
+// and output format
+type ExecConfig struct {
+	// Command to execute.
+	Command string `json:"command"`
+	// Arguments to pass to the command when executing it.
+	// +optional
+	Args []string `json:"args,omitempty"`
```

Comment:
Sorry if this was addressed in the KEP, but for beta I wonder if the exec args should be standardized across all providers (similar to CNI) for the request/response so there's no need to specify it in the provider config. The only parameters required would be the binary name as seen by the kubelet.

Reply:
Same with the env vars below, should those be standardized or arbitrary per provider?

Reply:
There already exists some projects that could potentially be modified to be credential providers (i.e. adding a command to https://github.com/awslabs/amazon-ecr-credential-helper) so I wanted to leave a little bit of flexibility for the providers in adding subcommands, etc.  That being said, in the above example there are no existing subcommands so we can choose a standard (i.e. <binary-name> get-credentials).  Env vars, I think should be left for the provider, because for example, the AWS sdk can be configured using environment variables to specify profile, region, and many others, so I can imagine a case where someone would want the ability customize which environment variables are passed to the plugin.  WDYT?

Reply:
Hmm, let me noodle on that one. Not a blocker for alpha IMO but something we should discuss in v1.19.

Reply:
So for alpha, we standardize on the argument that is passed to the binary (i.e. `get-credentials`) and we leave envs as is (set in config)?  Meaning we remove Args from the API type.

Reply:
For existing tools I think (?) it should be reasonable to create wrapper scripts that can implement `get-credentials` using those tools. 

I wonder if something like this would be a simpler starting point:

```
apiVersion: registrycredentials.k8s.io/v1alpha1
kind: RegistryCredentialConfig
providers:
- name: aws-ecr
  imageMatchers:
  - "*.dkr.ecr.*.amazonaws.com"
  - "*.dkr.ecr.*.amazonaws.com.cn"
  extraArgs:
    key1: data1
    key2: data2
```

The `command` can be implied from the provider name (e.g. `aws-ecr`), and the extra args can either be passed into the plugin via env vars or stdin. `args` is implied to be `get-credentials`. Maybe it's too late in the release to be suggesting changes to the type so I'm okay to discuss this post code-freeze.

Reply:
sgtm

Reply:
cc @cheftako